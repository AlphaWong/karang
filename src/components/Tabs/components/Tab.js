import React from 'react';
import styled from 'styled-components';
import { bool, node, string, func } from 'prop-types';
import { darken } from 'polished';

import Button from 'components/Button';
import { orange as primaryColor, lightGray } from 'styles/colors';
import noop from 'utils/noop';

import { TabContainer } from '../style';

const TabButton = styled(Button).attrs({ variant: 'link' })`
  padding: 1.25em;
  color: ${({ selected }) => (selected ? primaryColor : lightGray)};

  text-shadow: ${({ selected }) =>
    selected
      ? `-0.75px 0px 0px ${primaryColor}, -0.75px 0px 0px ${primaryColor}`
      : 'none'};

  &:hover:enabled {
    color: ${primaryColor};
    text-decoration: none;
  }

  &:active:enabled {
    color: ${darken(0.2, primaryColor)};
    text-decoration: none;
  }

  &:focus:enabled {
    text-decoration: none;
  }
`;

const Tab = ({ children, selected, onTabChange, name }) => (
  <TabContainer selected={selected}>
    <TabButton selected={selected} onClick={() => onTabChange(name)}>
      {children}
    </TabButton>
  </TabContainer>
);

Tab.propTypes = {
  /** Children elements */
  children: node,
  /** Name of the tab */
  name: string.isRequired, // eslint-disable-line react/no-typos
  /** Callback function. Managed by TabBar */
  onTabChange: func,
  /** Display the selected state. Managed by TabBar */
  selected: bool,
};

Tab.defaultProps = {
  children: null,
  onTabChange: noop,
  selected: false,
};

export default Tab;
