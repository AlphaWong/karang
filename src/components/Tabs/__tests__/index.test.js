import React from 'react';
import { mount, shallow } from 'enzyme';

import Tab, { TabBar } from '../index';

describe('Tab', () => {
  describe('Snapshots', () => {
    it('Deselected', () => {
      const wrapper = shallow(<Tab name="foo">Foo</Tab>);
      expect(wrapper).toMatchSnapshot();
    });

    it('Selected', () => {
      const wrapper = shallow(
        <Tab name="foo" selected>
          Foo
        </Tab>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Functions', () => {
    it('trigger onTabChange when clicked', () => {
      const tabChangeHandler = jest.fn();
      const wrapper = mount(
        <Tab name="foo" onTabChange={tabChangeHandler}>
          Foo
        </Tab>
      );
      wrapper.find('button').simulate('click');
      expect(tabChangeHandler).toHaveBeenCalledTimes(1);
    });
  });
});

describe('TabBar', () => {
  describe('Snapshots', () => {
    it('No tabs', () => {
      const wrapper = shallow(<TabBar />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Some tabs', () => {
      const wrapper = shallow(
        <TabBar>
          <Tab name="foo">Foo</Tab>
          <Tab name="bar">Bar</Tab>
        </TabBar>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('Selected tabs', () => {
      const wrapper = shallow(
        <TabBar activeTab="foo">
          <Tab name="foo">Foo</Tab>
          <Tab name="bar">Bar</Tab>
        </TabBar>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('Non-tab children', () => {
      const wrapper = shallow(<TabBar>Not a tab at all.</TabBar>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Functions', () => {
    it('calls onTabChange when a tab is clicked', () => {
      const tabChangeHandler = jest.fn();
      const wrapper = mount(
        <TabBar activeTab="foo" onTabChange={tabChangeHandler}>
          <Tab name="foo">Foo</Tab>
          <Tab name="bar">Bar</Tab>
        </TabBar>
      );

      wrapper
        .find('Tab[name="bar"]')
        .find('button')
        .simulate('click');
      expect(tabChangeHandler).toHaveBeenCalledWith('bar');
    });
  });
});
