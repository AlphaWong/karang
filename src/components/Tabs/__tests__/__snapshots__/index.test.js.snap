// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Snapshots Deselected 1`] = `
<styled.li
  selected={false}
>
  <Styled(Component)
    onClick={[Function]}
    selected={false}
  >
    Foo
  </Styled(Component)>
</styled.li>
`;

exports[`Tab Snapshots Selected 1`] = `
<styled.li
  selected={true}
>
  <Styled(Component)
    onClick={[Function]}
    selected={true}
  >
    Foo
  </Styled(Component)>
</styled.li>
`;

exports[`TabBar Snapshots No tabs 1`] = `""`;

exports[`TabBar Snapshots Non-tab children 1`] = `
<styled.ul>
  Not a tab at all.
</styled.ul>
`;

exports[`TabBar Snapshots Selected tabs 1`] = `
<styled.ul>
  <Tab
    key=".0"
    name="foo"
    onTabChange={[Function]}
    selected={true}
  >
    Foo
  </Tab>
  <Tab
    key=".1"
    name="bar"
    onTabChange={[Function]}
    selected={false}
  >
    Bar
  </Tab>
</styled.ul>
`;

exports[`TabBar Snapshots Some tabs 1`] = `
<styled.ul>
  <Tab
    key=".0"
    name="foo"
    onTabChange={[Function]}
    selected={false}
  >
    Foo
  </Tab>
  <Tab
    key=".1"
    name="bar"
    onTabChange={[Function]}
    selected={false}
  >
    Bar
  </Tab>
</styled.ul>
`;
